pkg/cloud/services/gc/cleanup.go

```diff
@@ -0,0 +1,109 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gc
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	rgapi "github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1beta1"
+	expinfrav1 "sigs.k8s.io/cluster-api-provider-aws/exp/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/annotations"
+)
+
+const (
+	serviceNameTag    = "kubernetes.io/service-name"
+	eksClusterNameTag = "aws:eks:cluster-name"
+)
+
+// ReconcileDelete is responsible for determining if the infra cluster needs to be garbage collected. If
+// does then it will perform garbage collection. For example, it will delete the ELB/NLBs that where created
+// as a result of Services of type load balancer.
+func (s *Service) ReconcileDelete(ctx context.Context) error {
+	s.scope.Info("reconciling deletion for garbage collection")
+
+	val, found := annotations.Get(s.scope.InfraCluster(), expinfrav1.ExternalResourceGCAnnotation)
+	if !found {
+		val = "true"
+	}
+
+	shouldGC, err := strconv.ParseBool(val)
+	if err != nil {
+		return fmt.Errorf("converting value %s of annotation %s to bool: %w", val, expinfrav1.ExternalResourceGCAnnotation, err)
+	}
+
+	if !shouldGC {
+		s.scope.Info("cluster opted-out of garbage collection")
+
+		return nil
+	}
+
+	return s.deleteResources(ctx)
+}
+
+func (s *Service) deleteResources(ctx context.Context) error {
+	s.scope.Info("deleting aws resources created by tenant cluster")
+
+	serviceTag := infrav1.ClusterAWSCloudProviderTagKey(s.scope.KubernetesClusterName())
+	awsInput := rgapi.GetResourcesInput{
+		ResourceTypeFilters: nil,
+		TagFilters: []*rgapi.TagFilter{
+			{
+				Key:    aws.String(serviceTag),
+				Values: []*string{aws.String(string(infrav1.ResourceLifecycleOwned))},
+			},
+		},
+	}
+
+	awsOutput, err := s.resourceTaggingClient.GetResourcesWithContext(ctx, &awsInput)
+	if err != nil {
+		return fmt.Errorf("getting tagged resources: %w", err)
+	}
+
+	resources := []*AWSResource{}
+
+	for i := range awsOutput.ResourceTagMappingList {
+		mapping := awsOutput.ResourceTagMappingList[i]
+		parsedArn, err := arn.Parse(*mapping.ResourceARN)
+		if err != nil {
+			return fmt.Errorf("parsing resource arn %s: %w", *mapping.ResourceARN, err)
+		}
+
+		tags := map[string]string{}
+		for _, rgTag := range mapping.Tags {
+			tags[*rgTag.Key] = *rgTag.Value
+		}
+
+		resources = append(resources, &AWSResource{
+			ARN:  &parsedArn,
+			Tags: tags,
+		})
+	}
+
+	for _, cleanupFunc := range s.cleanupFuncs {
+		if deleteErr := cleanupFunc(ctx, resources); deleteErr != nil {
+			return fmt.Errorf("deleting resources: %w", deleteErr)
+		}
+	}
```

Comment:
Optionally, this could be simplified here like this. Take the `s.cleanupFuncs` from this and have the type ( the slice type ) implement an interface which says `Execute(resources)`. That would internally loop on all its elements and call them with the same resources slice so we don't have to pass that slice around as many times as there are cleanup functions.

Reply:
I like this :+1:

Reply:
Changed.

Reply:
A receiver on a type alias of a slice. Very neat! I don't think I've seen this before, so it looks a little exotic to my eyes, but it's only one step removed from a struct that has a slice field.